### ISSUE TITLE
Variable named `exports` is not type-checked

### ISSUE THREAD

**Issue by guilhermesimoes** [microsoft/TypeScript#52197]
# Bug Report

### 🕗 Version & Regression Information

This is the behavior in every version I tried: v3.3.3 and v4.9.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/DYUwLgBCAeAOD2AnMBnAXBAdgVwLYCMREBuAKBgWRQgF4IByes00SAQwxwKLLdoaakgA)

### 💻 Code

```ts
let a: number;
a = ''; // Type 'string' is not assignable to type 'number'.

let exports: number;
exports = ''; // Everything is fine
```

### 🙁 Actual behavior

Variable named `exports` is not type-checked.

### 🙂 Expected behavior

Either the variable `exports` should be type-checked or the variable should not be allowed to be named `exports` (as far as I could tell `exports` is not a reserved word).

**Comment by Tobeyyyyy** [comment]
It seems like this behaviour was introduced back in [#23319](https://github.com/microsoft/TypeScript/pull/23319) to support assignment to `exports`. If the identifier's text is just `"exports"`, type checking is skipped. 

[checker.ts](https://github.com/microsoft/TypeScript/blob/05dc981ebfc7c86ff6c409d51639b2156bbc5f45/src/compiler/checker.ts#L35924) line 35924

```typescript
if (checkReferenceExpression(left,
  Diagnostics.The_left_hand_side_of_an_assignment_expression_must_be_a_variable_or_a_property_access,
  Diagnostics.The_left_hand_side_of_an_assignment_expression_may_not_be_an_optional_property_access)
  && (!isIdentifier(left) || unescapeLeadingUnderscores(left.escapedText) !== "exports")) {
  /* ... checkTypeAssignableToAndOptionallyElaborate ... */
}        
```

Now I am not sure if this check `!isIdentifier(left) || unescapeLeadingUnderscores(left.escapedText) !== "exports")` is needed anymore. I removed it in this [PR](https://github.com/microsoft/TypeScript/pull/52209) and the tests seem to pass.

**Comment by guilhermesimoes** [comment]
That was fast!

Fyi of how this came about. The implementation of [PeacockTV](https://www.peacocktv.com/) and [SkyShowtime](https://www.skyshowtime.com/) uses the same codebase for multiple platforms. We have a folder structure like this:
```
a-component
├── index.ts
├── index.androidtv.ts
├── index.xboxone.ts
├── etc
└── index.universal.ts
```
Builds for a specific platform resolve to that platform file instead of the default `index.ts` through the usage of [Webpack's resolve extensions](https://webpack.js.org/configuration/resolve/#resolveextensions).

This has one drawback which is that if an `index.androidtv.ts` export is different from an `index.ts` export then that's not type-checked. `tsc`'s module resolution is different from `Webpack`'s (since we configured it that way).

However, we have a "universal" build that imports code from all platforms and that at runtime can switch between them. This is not intended for production but is useful for testing on multiple platforms. Here's an example:
```ts
import { Platform } from './shared/platform';
import * as defaultExports from './index';
import * as androidtvExports from './index.androidtv';
import * as xboxoneExports from './index.xboxone';

let exports: typeof defaultExports;

switch (self.platform) {
  case Platform.AndroidTv:
    exports = androidtvExports;
    break;

  case Platform.XboxOne:
    exports = xboxoneExports;
    break;

  // etc

  default:
    exports = defaultExports;
}

export const initialiseDevice = exports.initialiseDevice;
export const getDeviceCapabilities = exports.getDeviceCapabilities;
```
So I was looking at this code and started to wonder "wait, why the hell don't we have type-safety on the `exports` assignments inside the switch?". And that's when I started to play with that variable and figured it wasn't checked at all.

In the mean time, we'll probably rename the variables `exports` on our side to fix any eventual incompatibilities we find so that we are ready for a TS version that fixes this.
